### ISSUE TITLE
rustc should warn about unused traits.

### ISSUE THREAD

**Issue by ehiggs** [rust-lang/rust#41883]
```rust
trait A{}

fn main() {}
```

This won't yield a compiler warning. However the equivalent with a struct will:

```rust
struct A{}

fn main() {}
```

https://play.rust-lang.org/?gist=0787e43a11bb28969e9f4978954c3fdc

rustc 1.19.0-nightly (d3abc80b3 2017-05-09)

**Comment by ehiggs** [comment]
I took a stab at fixing this:

```
diff --git a/src/librustc/middle/dead.rs b/src/librustc/middle/dead.rs
index 84ead6506c..1d93434840 100644
--- a/src/librustc/middle/dead.rs
+++ b/src/librustc/middle/dead.rs
@@ -430,6 +430,7 @@ impl<'a, 'tcx> DeadVisitor<'a, 'tcx> {
             | hir::ItemTy(..)
             | hir::ItemEnum(..)
             | hir::ItemStruct(..)
+            | hir::ItemTrait(..)
             | hir::ItemUnion(..) => true,
             _ => false
         };
```

But I can't rebuild rustc because:

```
   Compiling std v0.0.0 (file:///Users/ehiggs/src/download/rust/src/libstd)
error: trait is never used: `ZipImpl`
   --> src/libcore/iter/mod.rs:763:1
    |
763 | / trait ZipImpl<A, B> {
764 | |     type Item;
765 | |     fn new(a: A, b: B) -> Self;
766 | |     fn next(&mut self) -> Option<Self::Item>;
...   |
770 | |               B: DoubleEndedIterator + ExactSizeIterator;
771 | | }
    | |_^
    |
    = note: #[deny(dead_code)] implied by #[deny(warnings)]
note: lint level defined here
   --> src/libcore/lib.rs:68:9
    |
68  | #![deny(warnings)]
    |         ^^^^^^^^

error: trait is never used: `PointerExt`
    --> src/libcore/slice/mod.rs:1588:1
     |
1588 | / trait PointerExt : Copy {
1589 | |     unsafe fn slice_offset(self, i: isize) -> Self;
1590 | |
1591 | |     /// Increments `self` by 1, but returns the old value.
...    |
1604 | |     }
1605 | | }
     | |_^
     |
     = note: #[deny(dead_code)] implied by #[deny(warnings)]

error: trait is never used: `SplitIter`
    --> src/libcore/slice/mod.rs:1624:1
     |
1624 | / trait SplitIter: DoubleEndedIterator {
1625 | |     /// Marks the underlying iterator as complete, extracting the remaining
1626 | |     /// portion of the slice.
1627 | |     fn finish(&mut self) -> Option<Self::Item>;
1628 | | }
     | |_^
     |
     = note: #[deny(dead_code)] implied by #[deny(warnings)]

error: trait is never used: `SlicePartialEq`
    --> src/libcore/slice/mod.rs:2506:1
     |
2506 | / trait SlicePartialEq<B> {
2507 | |     fn equal(&self, other: &[B]) -> bool;
2508 | |
2509 | |     fn not_equal(&self, other: &[B]) -> bool { !self.equal(other) }
2510 | | }
     | |_^
     |
     = note: #[deny(dead_code)] implied by #[deny(warnings)]

error: trait is never used: `SlicePartialOrd`
    --> src/libcore/slice/mod.rs:2552:1
     |
2552 | / trait SlicePartialOrd<B> {
2553 | |     fn partial_compare(&self, other: &[B]) -> Option<Ordering>;
2554 | | }
     | |_^
     |
     = note: #[deny(dead_code)] implied by #[deny(warnings)]

error: trait is never used: `SliceOrd`
    --> src/libcore/slice/mod.rs:2588:1
     |
2588 | / trait SliceOrd<B> {
2589 | |     fn compare(&self, other: &[B]) -> Ordering;
2590 | | }
     | |_^
     |
     = note: #[deny(dead_code)] implied by #[deny(warnings)]

error: trait is never used: `BytewiseEquality`
    --> src/libcore/slice/mod.rs:2636:1
     |
2636 | trait BytewiseEquality { }
     | ^^^^^^^^^^^^^^^^^^^^^^^^^^
     |
     = note: #[deny(dead_code)] implied by #[deny(warnings)]

error: trait is never used: `GenericRadix`
   --> src/libcore/fmt/num.rs:51:1
    |
51  | / trait GenericRadix {
52  | |     /// The number of digits.
53  | |     fn base(&self) -> u8;
54  | |
...   |
100 | |     }
101 | | }
    | |_^
    |
    = note: #[deny(dead_code)] implied by #[deny(warnings)]

error: trait is never used: `TrustedRandomAccess`
  --> src/libcore/iter_private.rs:23:1
   |
23 | / pub unsafe trait TrustedRandomAccess : ExactSizeIterator {
24 | |     unsafe fn get_unchecked(&mut self, i: usize) -> Self::Item;
25 | |     /// Returns `true` if getting an iterator element may have
26 | |     /// side effects. Remember to take inner iterators into account.
27 | |     fn may_have_side_effect() -> bool;
28 | | }
   | |_^
   |
   = note: #[deny(dead_code)] implied by #[deny(warnings)]

error: aborting due to 9 previous errors

error: Could not compile `core`.
```

**Comment by mu001999** [comment]
@shepmaster This can be closed by #118257
